File: 백준/Gold/13023. ABCDE/ABCDE.py
import sys
sys.setrecursionlimit(10 ** 6)
input = sys.stdin.readline

N, M = map(int, input().split())
A = [[] for _ in range(N)]
visited = [False] * N

arrive = False

def DFS(v, depth):
    global arrive

    if depth == 5:
        arrive = True
        return
    
    visited[v] = True
    for i in A[v]:
        if not visited[i]:
            DFS(i, depth + 1)

    visited[v] = False

for _ in range(M):
    a, b = map(int, input().split())

    A[a].append(b)
    A[b].append(a)

for i in range(N):
    if not visited[i]:
        DFS(i, 1)
    if arrive:
        break
    
if arrive:
    print(1)
else:
    print(0)

File: 백준/Gold/13023. ABCDE/ABCDE.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.StringTokenizer;

public class Main {

    static ArrayList<Integer>[] A;
    static boolean[] visited;
    static boolean result;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int N = Integer.parseInt(st.nextToken());
        int M = Integer.parseInt(st.nextToken());

        A = new ArrayList[N];
        visited = new boolean[N];

        for (int i = 0; i < N; i++) {
            A[i] = new ArrayList<>();
        }

        for (int i = 0; i < M; i++) {
            st = new StringTokenizer(br.readLine());

            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());

            A[a].add(b);
            A[b].add(a);
        }

        result = false;
        for (int i = 0; i < N; i++) {
            if (result)
                break;

            DFS(i, 1);

//            for (int j = 0; j < N; j++) {
//                visited[j] = false;
//            }
        }

        if (result) System.out.println(1);
        else System.out.println(0);
    }

    static void DFS(int v, int depth) {
        if (depth == 5 || result) {
            result = true;
            return;
        }

        visited[v] = true;
        for (int i : A[v]) {
            if (!visited[i]) {
                DFS(i, depth + 1);
            }
        }
        visited[v] = false;
    }
}
